lib/scholar/naive_bayes/gaussian_naive_bayes.ex

```diff
@@ -0,0 +1,422 @@
+            [1.1250000042650754e-8, 1.1250000042650754e-8, 1.1250000042650754e-8],
+            [1.1250000042650754e-8, 1.1250000042650754e-8, 1.1250000042650754e-8],
+            [9.0, 9.0, 9.0]
+          ]
+        ),
+        class_count: Nx.tensor([1.0, 1.0, 2.0]),
+        class_priors: Nx.tensor([0.25, 0.25, 0.5]),
+        classes: Nx.tensor([0, 1, 2]),
+        epsilon: Nx.tensor(1.1250000042650754e-8)
+      }
+
+      iex> x = Nx.iota({4, 3})
+      iex> y = Nx.tensor([1, 2, 0, 2])
+      iex> Scholar.NaiveBayes.Gaussian.fit(x, y, num_classes: 3, sample_weights: [1, 6, 2, 3])
+      %Scholar.NaiveBayes.Gaussian{
+        theta: Nx.tensor(
+          [
+            [6.0, 7.0, 8.0],
+            [0.0, 1.0, 2.0],
+            [5.0, 6.0, 7.0]
+          ]
+        ),
+        var: Nx.tensor(
+          [
+            [1.1250000042650754e-8, 1.1250000042650754e-8, 1.1250000042650754e-8],
+            [1.1250000042650754e-8, 1.1250000042650754e-8, 1.1250000042650754e-8],
+            [8.0, 8.0, 8.0]
+          ]
+          ),
+        class_count: Nx.tensor([2.0, 1.0, 9.0]),
+        class_priors: Nx.tensor([0.1666666716337204, 0.0833333358168602, 0.75]),
+        classes: Nx.tensor([0, 1, 2]),
+        epsilon: Nx.tensor(1.1250000042650754e-8)
+      }
+  """
+  deftransform fit(x, y, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    input_rank = Nx.rank(x)
+    targets_rank = Nx.rank(y)
+
+    if input_rank != 2 or targets_rank != 1 do
+      raise ArgumentError,
+            "Wrong input rank. Expected: 2 for x and 1 for y, got: #{inspect(input_rank)} for x and #{inspect(targets_rank)} for y"
+    end
+
+    {num_samples, _} = Nx.shape(x)
+    {num_targets} = Nx.shape(y)
+
+    if num_samples != num_targets do
+      raise ArgumentError,
+            "Wrong input shape. Expect x to have the same first dimension as y, got: #{inspect(num_samples)} for x and #{inspect(num_targets)} for y"
+    end
+
+    opts =
+      opts ++
+        [
+          sample_weights_flag: if(opts[:sample_weights] != nil, do: true, else: false),
+          priors_flag: if(opts[:priors] != nil, do: true, else: false)
+        ]
+
+    sample_weights =
+      case opts[:sample_weights_flag] do
+        false -> Nx.broadcast(1.0, {num_samples})
+        _ -> Nx.tensor(opts[:sample_weights])
+      end
+
+    priors = opts[:priors]
+
+    classes = Nx.iota({opts[:num_classes]})
+
+    class_priors =
+      case opts[:priors_flag] do
+        false ->
+          Nx.broadcast(0.0, Nx.shape(classes))
+
+        _ ->
+          priors = Nx.tensor(priors)
+
+          if Nx.size(priors) != Nx.size(classes) do
+            raise ArgumentError,
+                  "Number of priors must match number of classes. Number of priors: #{inspect(Nx.size(priors))} does not match number of classes: #{inspect(Nx.size(classes))}"
+          else
+            priors
+          end
+      end
+
+    fit_n(x, y, sample_weights, class_priors, classes, opts)
```

Comment:
Sorry, this is a little bit annoying to handle, but you want to call as few tensor operations as possible outside of `defn`. Usually, you want to leave only tensor allocations outside:

```
    # Remove the value from opts so it is not part of the cache
    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, 1.0)
    sample_weights = Nx.tensor(sample_weights)

    {priors, opts} = Keyword.pop(opts, :priors, 0.0)
    priors = Nx.tensor(priors)

    fit_n(x, y, sample_weights, class_priors, opts)
```

And move all broadcasting, iota, etc in:

```
    num_classes = opts[:num_classes]
    classes = Nx.iota({num_classes})

    priors =
      case Nx.shape(priors) do
        {} -> Nx.broadcast(priors, {num_classes})
        {^num_classes} -> priors
        _ -> raise ArgumentError, "..."
      end

    {num_samples, _} = Nx.shape(x)

    sample_weights =
      case Nx.shape(sample_weights) do
        {} -> Nx.broadcast(sample_weights, {num_samples})
        {^num_samples} -> sample_weights
        _ -> raise ArgumentError, "..."
      end
```

Does it make sense?

Reply:
Ok, sure